README.md
Adafruit webIDE
================
This is a simple editor designed to help learn the Raspberry Pi and Beaglebone components, and more. This editor is designed solely for use on your secure private network as of now.

Debian Installation (Raspberry Pi and BeagleBone Black)
============

The WebIDE installer is currently targeting Debian Stretch (latest Raspbian) installations only.

On the Raspberry PI or BeagleBone Black (after expanding the file system):

    curl https://raw.githubusercontent.com/adafruit/Adafruit-WebIDE/master/scripts/install.sh | sudo sh

Alternatively, you can install using the .deb file:

    curl -O https://adafruit-download.s3.amazonaws.com/adafruitwebide-0.3.12-Linux.deb
    sudo dpkg -i adafruitwebide-0.3.12-Linux.deb
    sudo apt-get -f install

Note: As part of the installation process, the 'webide' user is given access to sudo and sudoers,
similar to the 'pi' user.  This is needed in order to easily access GPIO pins from the Editor.  
If you don't need these features, feel free to manually install the editor below.

Note: This is also the default installation for any Debian or Ubuntu operating systems

Manual Installation
============

Follow along in the [installation script][1] and pick and choose
the components you'd like to install.

Uninstallation
============

Debian (Raspberry PI and BeagleBone Black):

    curl https://raw.githubusercontent.com/adafruit/Adafruit-WebIDE/master/scripts/uninstall.sh | sudo sh

Manual Uninstallation
============

Follow along in the [uninstallation script][2] and pick and choose
the components you'd like to remove.

Running the Editor
============

Using Firefox or Chrome (and likely any other webkit browser) on any computer in your internal network:

Raspberry Pi:

    http://raspberrypi.local:8080

BeagleBone:

    http://beaglebone.local:8080

Restart the Editor
============

If for any reason you need to restart the editor, you can execute the following commands in order

    sudo systemctl restart adafruit-webide
    sudo systemctl start adafruit-webide

Sudo is required to restart due to the editor running as the 'webide' user.

Status or Logs for the Editor
============

sudo systemctl status adafruit-webide

Logs are in syslog: /var/log/syslog

License
============

The editor is licensed with AGPL Version 3.
http://www.gnu.org/licenses/agpl-3.0.html

SCREENSHOTS
===========
![ScreenShot](http://www.adafruit.com/adablog/wp-content/uploads/2012/10/WebIDE_Alpha.jpg)

[1]: https://github.com/adafruit/Adafruit-WebIDE/blob/master/scripts/install.sh
[2]: https://github.com/adafruit/Adafruit-WebIDE/blob/master/scripts/uninstall.sh


release/changelog.txt
-0.8.0
Add simple python3 support
Fix debug populating variables

-0.7.0
Update pug dependency
Updates to installer and uninstaller

-0.6.0
Initial pre-release of the simplified webide

-0.3.12
Add missing cookie package

-0.3.11
Fixes for installation errors
Fix some deprecated messages

-0.3.10
Fix for settings being undefined in offline mode

-0.3.9
Include support for Angstrom (Beaglebone)
Re-brand as the Adafruit webIDE

-0.3.8
Update README with better manual installation instructions (Thanks adamboutcher!)
Update tty.js dependencies for issue #178
Add more transports for socket.io to resolve websocket issues.
Fix for renaming file selection deleting unsaved editor file contents.
Handle file rename errors, including detecting renames with same name.

-0.3.7
Github support.
 - Ability to enable with --github as the default
 - Advanced setting.
 - Requires manual ssh key setup as of yet.
 - Most commands are treated as manual mode for now (manual commits, etc).
New Offline mode
 - Ability to install with --offline as the default
 - need to manually commit, and push changes (similar to the manual git setting)
 - Bypasses bitbucket OAuth
Ability to refresh directories from within the navigator
New option to clone repositories without updating remote to bitbucket
New right-click context menu option to update repositories from remote (origin/master for now)
New Report Bug Link added to footer
New confirmation dialog for navigating away from unsaved changes...Save Files/Don't Save/Cancel
Editor is set to readOnly for any files that shouldn't change (README, update notes), including empty editor window while navigating.
Deleting a file or project will now also delete a corresponding scheduled job from the queue.
Errors cloning repositories are sent to the front-end now.
Error handling for most git commands now.  Notifications visible in webide for failures.
New Error pages for any issues with the system failing to show pages.  Links to ALS WebIDE FAQ for help.
New Error page specifically for OAuth failures.  Adds a button to execute a script to help set the date and time.
Attempt to set the date on the Pi during installation to prevent OAuth errors.
Creating files and folders will automatically open them in the editor and navigator.
Uploaded files will always use the current working directory, instead of uploading to the parent directory now.
git pull commands are now using the quiet (-q) flag.
Editor setting for supporting adding a Make link in the editor action bar if a Makefile is detected in the cwd.  Not enabled by default.

-0.3.6
Update Font Awesome to latest version
Fix for git commit and push when settings are already configured
Modify socket.io for better performance

-0.3.5
Fix crashing on startup of new installations.

-0.3.4
Fix for code not getting pushed out to remote repository.
List folders and files alphabetically.
Remove long running call causing slowdowns.

-0.3.3
Add four new settings to the Settings Page
  -Use Soft Tabs (on or off)
  -Tab Size (2, 4, 6, or 8)
  -Show Invisibles (on or off) (shows whitespaces, tabs and spaces, etc)
  -Manual Git Commits and Pushes
Add ability to manually commit and push file changes.  This includes adding comments.
  -Do a git status on the file to check if it is in need of a commit and push, and populate editor bar.
Set editor defaults to use soft tabs with a setting of 4 tab size, and show invisibles off
Add a default .gitignore file to reduce chance of .pyc files getting pushed out to bitbucket.
Fix for the terminal getting out of sync with the editor
Prevent overwriting of Adafruit project files
Update manual installation instructions
Install node.js as part of installation again.

-0.3.2
Navigator: Add ability to create folders at all levels.
Debugger: Fix display of local vs global variables.
Terminal: Set terminal width based on containing parent, use entire width and height.
Terminal: Resize browser window resizes terminal appropriately.
Remove file creation and uploads in adafruit repository (it would fail anyways) project

-0.3.1
Fix for debugger hanging when stderr is written from script being debug.

-0.3.0
New Python Debugger
Upgrade Node.js to 0.8 (new installations only)
Include Node.js binaries with WebIDE
Fix Cloning repositories originating from Bitbucket.

-0.2.5
Run visualizer as sudo to allow access to GPIO

-0.2.4
New visualization feature for Python
Added ability to change font sizes in the editor and the terminal
New settings manager added, currently only has font size, but more can be added
Fix for long file names causing strange wrapping

-0.2.3
Fix server failing to start on reboot of Pi.

-0.2.2
Fix for false redirects to /setup even if bitbucket and git is already configured.
Change page title based on what directory or file is open.
Update Ace to the latest version, fixing line numbering
Detect if port 80 is in use during install, and try 3000 instead.
Set path to open editor at end of install script to actual hostname.
Install python-smbus as part of the installation process.
Add symlink for repositories to webide users home path.
Fix for closing the schedule manager breaking the editor toolbar.
Set default as using hard tabs for editing files for now.

-0.2.1
Job scheduler for automating script execution.  Run a script "every 15 minutes", for example.
Schedule manager for toggling jobs, and deleting jobs.
Move terminal inline with the editor for faster run/code workflow.
Focus editor when running code.
Notifications while running scripts in editor bar.
Focus the editor in many more situations, for example, when opening a file, it will give focus to the editor.
Add "Command-Enter", or "Ctrl-Enter" command to run scripts, this is in addition to "Command-S" and "Ctrl-S" to save.
Update Bitbucket logos with new design.
Add option to change port number in /config, and persist in Redis, surviving updates.
Ensure only one instance of restartd is running during installation

config/config.js
//NOTE: Changes in this file will not persist between updates at this point.
//Change port number in /config.
//Change offline in /config.
//Change github in /config.
exports.editor = {
  "port": 3000,
  "version": "0.8.0",
  "version_url": "https://raw.githubusercontent.com/adafruit/Adafruit-WebIDE/master/release/version.txt"
};

exports.adafruit = {
  "repository": "Adafruit-Raspberry-Pi-Python-Code",
  "remote": "git://github.com/adafruit/Adafruit-Raspberry-Pi-Python-Code.git",
  "remote_name": "adaremote"
};

exports.defaults = {
  "repository": "my-pi-projects",
  "readme": "README.md",
  "gitignore": ".gitignore"
};

exports.term = {
  "shell": "bash",
  "cwd": "./repositories",
  "log": true,
  "term": {
    "termName": "xterm",
    "geometry": [80, 15],
    "scrollback": 1000,
    "visualBell": false,
    "popOnBell": false,
    "cursorBlink": false,
    "screenKeys": false,
    "colors": [
      "#2e3436",
      "#cc0000",
      "#4e9a06",
      "#c4a000",
      "#3465a4",
      "#75507b",
      "#06989a",
      "#d3d7cf",
      "#555753",
      "#ef2929",
      "#8ae234",
      "#fce94f",
      "#729fcf",
      "#ad7fa8",
      "#34e2e2",
      "#eeeeec"
    ]
  }
};
